File: repositories/langtags/setting_repository.py
from core.sqlite.standard_table import StandardTable
from .language_repository import LanguageRepository

class SettingRepository():
    def __init__(self, table: StandardTable ):

        self.table = table
        self.database = self.table.database
        self._COLUMN_ID = 'setting_id'
        self._COLUMN_PARAMETER = 'parameter_name'
        self._COLUMN_LANGUAGE_ID = "language_id"

    def update_parameter_language_id(self, parameter_name:str, language_id:int):
        try:
            cursor = self.database.execute(
                statement=(
                    f"UPDATE {self.table.name} SET {self._COLUMN_LANGUAGE_ID}=? WHERE {self._COLUMN_PARAMETER}=?;"
                ),  commit=True, params=( language_id, parameter_name )
            )
            return True
        except:
            return False

    def insert_parameter_language_id(self, parameter_name:str, language_id:int):
        try:
            cursor = self.database.execute(
                statement=(
                    f"INSERT INTO {self.table.name} ({self._COLUMN_PARAMETER}, {self._COLUMN_LANGUAGE_ID}) VALUES(?, ?);"
                ),
                commit=True, params=( parameter_name, language_id )
            )
            return True
        except:
            return False

    def update_parameter_language_code(self, parameter_name, language_code):
        try:
            cursor = self.database.execute(
                statement=(
                    f"UPDATE {self.table.name} SET {self._COLUMN_LANGUAGE_ID} = "
                    f"(SELECT l.language_id FROM languages l "
                    f"WHERE l.code=? AND l.deleted_at IS NULL) "
                    f"WHERE {self._COLUMN_PARAMETER}=?;"
                ), commit=True, params=(language_code, parameter_name)
            )
            return True
        except:
            return False

    def update_current_language_id(self, language_id:int ):
        return self.update_parameter( 'current_language', language_id )

    def update_default_language_id(self, language_id:int ):
        return self.update_parameter( 'default_language', language_id )

    def update_system_language_id(self, language_id:int ):
        return self.update_parameter( 'system_language', language_id )

    def update_current_language_code(self, language_code:str ):
        return self.update_parameter_code( 'current_language', language_code )

    def update_default_language_code(self, language_code:str ):
        return self.update_parameter_code( 'default_language', language_code )

    def update_system_language_code(self, language_code:str ):
        return self.update_parameter_code( 'system_language', language_code )

    def select_parameter_language_id(self, parameter_name:str) -> int | None:
        try:
            cursor = self.database.execute(
                statement=(
                    f"SELECT {self._COLUMN_LANGUAGE_ID} FROM {self.table.name} WHERE " f"{self._COLUMN_PARAMETER}=?"
                ), commit=False, params=(parameter_name,),
            )
            value = cursor.fetchone()
            return value[0]
        except:
            return None

    def select_parameter_language_code(self, parameter_name) -> str | None:
        try:
            cursor = self.database.execute(
                statement=(
                    f"SELECT l.code FROM {self.table.name} s JOIN languages l "
                    f"ON s.{self._COLUMN_LANGUAGE_ID}=l.language_id "
                    f"WHERE s.{self._COLUMN_PARAMETER}=? AND l.deleted_at IS NULL"
                ), commit=False, params=(parameter_name,)
            )
            value = cursor.fetchone()
            return value[0]
        except:
            return None

    def select_current_language_id(self):
        return self.select_parameter_language_id( "current_language" )

    def select_default_language_id(self):
        return self.select_parameter_language_id( "default_language" )

    def select_system_language_id(self):
        return self.select_parameter_language_id( "system_language" )

    def select_current_language_code(self):
        return self.select_parameter_language_code( "current_language" )

    def select_default_language_code(self):
        return self.select_parameter_language_code( "default_language" )

    def is_current_language_system(self) -> bool:
        current_id = self.select_current_language_id()
        system_id = self.select_system_language_id()
        return current_id == system_id

    def is_current_language_default(self) -> bool:
        current_id = self.select_current_language_id()
        default_id = self.select_default_language_id()
        return current_id == default_id

File: Repositories/LangTags/SettingRepository.cs
// Core 
using Core;
using Core.Sqlite;

namespace Repositories.LangTags
{
    public class SettingRepository
    {
        // Variables
        private StandardDatabase _db;
        private StandardTable _table;
        private LanguageRepository _languageRepository; 
        
        // Constructor
        public SettingRepository(StandardDatabase standardDatabase, LanguageRepository languageRepository)
        {
            _db = standardDatabase;
            _table = new StandardTable(_db, "settings");
            _languageRepository = languageRepository;
        }

        public void InitParameters()
        {
            if (!ExistsByParameterName("default_language"))
            {
                Save("default_language", 1);
            }
            if (!ExistsByParameterName("selected_language"))
            {
                Save("selected_language", 1);
            }
            if (!ExistsByParameterName("system_language"))
            {
                Save("system_language", 0);
            }
        }

        // Propiedades
        public StandardTable Table { get{ return _table;} }

        // Methods
        public void Insert(string parameterName, int languageId)
        {
            _db.Execute(
                sql: "INSERT INTO settings (parameter_name, language_id) VALUES(@p0, @p1);", 
                commit:true, parameterName, languageId 
            );
        }
        public void Update(int settingId, string parameterName, int languageId)
        {
            _db.Execute(
                sql: "UPDATE settings SET language_id=@p0 WHERE parameter_name=@p1;", 
                commit:true, languageId, parameterName 
            );
        }

        public bool ExistsById(int settingId)
        {
            using var reader = _db.Query(
                sql: "SELECT 1 FROM settings WHERE setting_id=@p0 LIMIT 1;",
                settingId
            );
            return reader.Read();
        }
        public bool ExistsByParameterName(string parameterName)
        {
            using var reader = _db.Query(
                sql: "SELECT 1 FROM settings WHERE parameter_name=@p0 LIMIT 1;", 
                parameterName
            );
            return reader.Read();
        }

        public int GetIdByParameterName(string parameterName){
            using var reader = _db.Query(
                sql: "SELECT setting_id FROM settings WHERE parameter_name=@p0 LIMIT 1;",
                parameterName
            );
            reader.Read();
            return reader.GetInt32(0);
        }
        public string GetParameterNameById(int settingId){
            using var reader = _db.Query(
                sql: "SELECT parameter_name FROM settings WHERE setting_id=@p0 LIMIT 1;",
                settingId
            );
            reader.Read();
            return reader.GetString(0);
        }
        public int GetLanguageIdByParameterName(string parameterName){
            using var reader = _db.Query(
                sql: "SELECT language_id FROM settings WHERE parameter_name=@p0 LIMIT 1;",
                parameterName
            );
            reader.Read();
            return reader.GetInt32(0);
        }
        public string GetParameterNameByLanguageId(int languageId){
            using var reader = _db.Query(
                sql: "SELECT parameter_name FROM settings WHERE language_id=@p0 LIMIT 1;",
                languageId
            );
            reader.Read();
            return reader.GetString(0);
        }

        public void Save(string parameterName, int languageId, int? settingId=null)
        {
            // Determinar que exista
            bool exists = false;
            int id = 0;
            if (settingId == null) {
                exists = ExistsByParameterName(parameterName);
                if (exists){
                    id = GetIdByParameterName(parameterName);
                }
            } else {
                id = settingId.Value;
                exists = ExistsById(id);
            }
            
            // Escribir data.
            if (exists) {
                Update(id, parameterName, languageId);
            } else {
                Insert(parameterName, languageId);
            }
        }


        public int GetSelectedLanguageId()
        {
            return GetLanguageIdByParameterName( "selected_language" );
        }
        public string GetSelectedLanguageCode()
        {
            int id = GetSelectedLanguageId();
            if (id == GetSystemLanguageId())
            {
                return GetSystemLanguageCode();
            }
            else 
            {
                return _languageRepository.GetCodeById(id);
            }
        }
        public void UpdateSelectedLanguageId(int languageId)
        {
            int id = GetIdByParameterName("selected_language");
            Update(id, "selected_language", languageId);
        }
        public void UpdateSelectedLanguageCode(string code)
        {
            int id = _languageRepository.GetIdByCode(code);
            UpdateSelectedLanguageId(id);
        }

        public int GetDefaultLanguageId()
        {
            return GetLanguageIdByParameterName( "default_language" );
        }
        public string GetDefaultLanguageCode()
        {
            int id = GetDefaultLanguageId();
            return _languageRepository.GetCodeById(id);
        }
        public int GetSystemLanguageId()
        {
            return GetLanguageIdByParameterName( "system_language" );
        }
        public string GetSystemLanguageCode()
        {
            return SystemLanguage.GetCode();
        }

        public void EstablishDefaultLanguage()
        {
            UpdateSelectedLanguageId(GetDefaultLanguageId());
        }
        public void EstablishSystemLanguage()
        {
            UpdateSelectedLanguageId(GetSystemLanguageId());
        }
    }
}